src/VirtoCommerce.Platform.Data/GenericCrud/CrudService.cs

```diff
@@ -0,0 +1,207 @@
+    {
+        protected readonly IEventPublisher _eventPublisher;
+        protected readonly IPlatformMemoryCache _platformMemoryCache;
+        protected readonly Func<IRepository> _repositoryFactory;
+
+        protected CrudService(IEventPublisher eventPublisher, IPlatformMemoryCache platformMemoryCache, Func<IRepository> repositoryFactory)
+        {
+            _eventPublisher = eventPublisher;
+            _platformMemoryCache = platformMemoryCache;
+            _repositoryFactory = repositoryFactory;
+        }
+
+        public virtual async Task<TModel> GetByIdAsync(string id, string responseGroup = null)
+        {
+            var entities = await GetByIdsAsync(new[] { id }, responseGroup);
+            return entities.FirstOrDefault();
+        }
+
+        public virtual async Task<IEnumerable<TModel>> GetByIdsAsync(IEnumerable<string> ids, string responseGroup = null)
+        {
+            var cacheKey = CacheKey.With(GetType(), nameof(GetByIdsAsync), string.Join("-", ids), responseGroup);
+            var result = await _platformMemoryCache.GetOrCreateExclusiveAsync(cacheKey, async (cacheEntry) =>
+            {
+                var models = new List<TModel>();
+
+                using (var repository = _repositoryFactory())
+                {
+                    //Disable DBContext change tracking for better performance 
+                    repository.DisableChangesTracking();
+
+                    //It is so important to generate change tokens for all ids even for not existing objects to prevent an issue
+                    //with caching of empty results for non - existing objects that have the infinitive lifetime in the cache
+                    //and future unavailability to create objects with these ids.
+                    cacheEntry.AddExpirationToken(CreateCacheToken(ids));
+
+                    var entities = await LoadEntities(repository, ids, responseGroup);
+
+                    foreach (var entity in entities)
+                    {
+                        var model = entity.ToModel(AbstractTypeFactory<TModel>.TryCreateInstance());
+                        model = ProcessModel(responseGroup, entity, model);
+                        if (model != null) models.Add(model);
+                    }
+
+                }
+
+                return models;
+            });
+
+            return result.Select(x => (TModel)x.Clone());
+        }
+
+        protected virtual TModel ProcessModel(string responseGroup, TEntity entity, TModel model)
+        {
+            return model;
+        }
+
+        protected abstract Task<IEnumerable<TEntity>> LoadEntities(IRepository repository, IEnumerable<string> ids, string responseGroup);
+
+        protected Task<IEnumerable<TEntity>> LoadEntities(IRepository repository, IEnumerable<string> ids)
+        {
+            return LoadEntities(repository, ids, "Full");
+        }
+
+        protected virtual Task BeforeSaveChanges(IEnumerable<TModel> models)
```

Comment:
why do not make these methods abstract?

Reply:
Think, there are services without such actions